[quant] PerChannelFloatQParams support for quint4x2 dtype

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45594 [quant] PerChannelFloatQParams support for quint4x2 dtype**
* #44678 [quant] creating quint4x2 dtype for quantized tensors

Summary:
Adds support for Per-channel quantization using float qparams for 4-bit dtype
We use the new dispatch mechanism and use existing quantize/dequantize kernels to pack the
4-bit data depending on the bit_width.
Size of 4-bit quantized tensor is half that of 8-bit quantized tensor.

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_quantize_per_channel_sub_byte

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24025595](https://our.internmc.facebook.com/intern/diff/D24025595)

supriyar committed:
[quant] PerChannelFloatQParams support for quint4x2 dtype

Summary:
Adds support for Per-channel quantization using float qparams for 4-bit dtype
We use the new dispatch mechanism and use existing quantize/dequantize kernels to pack the
4-bit data depending on the bit_width.
Size of 4-bit quantized tensor is half that of 8-bit quantized tensor.

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_quantize_per_channel_sub_byte

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] PerChannelFloatQParams support for quint4x2 dtype"


Summary:
Adds support for Per-channel quantization using float qparams for 4-bit dtype
We use the new dispatch mechanism and use existing quantize/dequantize kernels to pack the
4-bit data depending on the bit_width.
Size of 4-bit quantized tensor is half that of 8-bit quantized tensor.

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_quantize_per_channel_sub_byte

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24025595](https://our.internmc.facebook.com/intern/diff/D24025595)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] PerChannelFloatQParams support for quint4x2 dtype"


Summary:
Adds support for Per-channel quantization using float qparams for 4-bit dtype
We use the new dispatch mechanism and use existing quantize/dequantize kernels to pack the
4-bit data depending on the bit_width.
Size of 4-bit quantized tensor is half that of 8-bit quantized tensor.

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_quantize_per_channel_sub_byte

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24025595](https://our.internmc.facebook.com/intern/diff/D24025595)

[ghstack-poisoned]

raghuramank100 reviewed (approved):
Does dequantize() work with no changes?

supriyar commented:
> Does dequantize() work with no changes?

I've made changes to the `dequantize_per_channel_affine_kernel` to do the unpacking when dequantizing.

supriyar committed:
Update on "[quant] PerChannelFloatQParams support for quint4x2 dtype"


Summary:
Adds support for Per-channel quantization using float qparams for 4-bit dtype
We use the new dispatch mechanism and use existing quantize/dequantize kernels to pack the
4-bit data depending on the bit_width.
Size of 4-bit quantized tensor is half that of 8-bit quantized tensor.

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_quantize_per_channel_sub_byte

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24025595](https://our.internmc.facebook.com/intern/diff/D24025595)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] PerChannelFloatQParams support for quint4x2 dtype"


Summary:
Adds support for Per-channel quantization using float qparams for 4-bit dtype
We use the new dispatch mechanism and use existing quantize/dequantize kernels to pack the
4-bit data depending on the bit_width.
Size of 4-bit quantized tensor is half that of 8-bit quantized tensor.

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_quantize_per_channel_sub_byte

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24025595](https://our.internmc.facebook.com/intern/diff/D24025595)

[ghstack-poisoned]
